marklogic-client-api/src/test/java/com/marklogic/client/test/datamovement/JSONSplitterTest.java

```diff
@@ -0,0 +1,241 @@
+            assertNotNull(docOp.getUri());
+
+            assertNotNull(docOp.getContent());
+            String docOpContent = docOp.getContent().toString();
+            assertEquals(docOpContent, expectedArray[i]);
+            i++;
+        }
+
+        assertEquals(5, splitter.getCount());
+
+    }
+
+    @Test
+    public void testCustomMultiArray() throws Exception {
+
+        MultiArrayVisitor visitor = new MultiArrayVisitor();
+        JSONSplitter splitter = new JSONSplitter(visitor);
+        FileInputStream fileInputStream = new FileInputStream(new File(jsonCustMultiArrayFile));
+        Stream<StringHandle> contentStream = splitter.split(fileInputStream);
+        assertNotNull(contentStream);
+
+        StringHandle[] result = contentStream.toArray(size -> new StringHandle[size]);
+        assertEquals(2, splitter.getCount());
+        assertNotNull(result);
+
+        String[] expected = new String[] {
+                "{\"record\":\"first record\"}",
+                "{\"record\":\"second record\"}"
+        };
+
+        for (int i = 0; i < result.length; i++) {
+            String element = result[i].get();
+            assertNotNull(element);
+            assertEquals(expected[i], element);
+        }
+    }
+
+    static public class MultiArrayVisitor extends JSONSplitter.Visitor {
+
+        @Override
+        public NodeOperation startArray(String containerKey) {
+            arrayDepth++;
+            return NodeOperation.DESCEND;
+        }
+
+        @Override
+        public StringHandle makeBufferedHandle(JsonParser containerParser) {
+            if (containerParser == null) {
+                throw new IllegalArgumentException("JsonParser cannot be null");
+            }
+            String content = serialize(containerParser);
+            return new StringHandle(content).withFormat(Format.JSON);
+        }
+    }
+
+    @Test
+    public void testCustomKey() throws Exception {
+
+        KeyVisitor visitor = new KeyVisitor("context2");
+        JSONSplitter splitter = new JSONSplitter(visitor);
+        FileInputStream fileInputStream = new FileInputStream(new File(jsonObjectFile));
+        Stream<StringHandle> contentStream = splitter.split(fileInputStream);
+        assertNotNull(contentStream);
+
+        StringHandle[] result = contentStream.toArray(size -> new StringHandle[size]);
+        assertEquals(3, splitter.getCount());
+        assertNotNull(result);
+
+        String[] expected = new String[] {
+                "{\"record\":\"third record\"}",
+                "[{\"record\":\"forth record\"},{\"record\":\"fifth record\"}]",
+                "{\"context\":[{\"record\":\"sixth record\"},{\"record\":\"seventh record\"}]}"
+        };
+
+        for (int i = 0; i < result.length; i++) {
+            String element = result[i].get();
+            assertNotNull(element);
+            assertEquals(expected[i], element);
+        }
+    }
+
+    static public class KeyVisitor extends JSONSplitter.Visitor {
```

Comment:
Should this test subclass ArrayVisitor if it needs to manage array depth?

Reply:
Yes, fixed as in the above comment.